spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/Session.scala

```diff
@@ -44,18 +47,364 @@ abstract class Session[T: ClassTag] {
+
+    val batchRdd = rdd.mapPartitions { iter =>
+
+      new Iterator[Table] {
+        override def hasNext: Boolean = iter.hasNext
+
+        override def next(): Table = {
+          require(iter.hasNext, "Call next() on a empty iterator")
+          val batch = for (_ <- 0 until batchSize if iter.hasNext) yield {
+            iter.next()
+          }
+          pack(batch)
+        }
+      }
+
+    }
+    batchRdd
+  }
+
+  private def pack(tables: Seq[Table], dimension: Int = 1): Table = {
+    val batch = tables.map(tableToSeq)
+    val firstSeq = batch.head
+    val sizes = firstSeq.map { tensor =>
+      val nDim = tensor.nDimension()
+      val size: Array[Int] = new Array[Int](nDim + 1)
+      var i = 1
+      while(i <= nDim + 1) {
+        if (i < dimension) {
+          size(i-1) = tensor.size(i)
+        } else if (i == dimension) {
+          size(i-1) = batch.length
+        } else {
+          size(i-1) = tensor.size(i - 1)
+        }
+        i = i + 1
+      }
+      size
+    }
+
+    val results = sizes.map { size =>
+      Tensor[T](size)
+    }
+
+    for ((seq, index) <- batch.zipWithIndex) {
+      results.zip(seq).foreach { case (result, tensor) =>
+        result.narrow(dimension, index + 1, 1).copy(tensor)
+      }
+    }
+    seqToTable(results)
+  }
+
+  type DataCache = mutable.HashMap[String, Array[Seq[Table]]]
+
+  private def adjustInputNames(inputs: Seq[String]): Seq[String] = {
+    val stripedNames = inputs.map(_.split(":")(0))
+    val set = mutable.LinkedHashSet[String]()
+    for (name <- stripedNames) {
+      set.add(name)
+    }
+    set.toSeq
+  }
+
+  def constructLocalData(endPoints: Seq[String], cache: DataCache): Seq[Table] = {
+    val isInputOp = (n: NodeDef) => inputOp(n.getOp)
+    val (tfGraph, inputs, originInputs) = TensorflowLoader.
+      buildTFGraph(graph.asJava, endPoints, isInputOp)
+
+    val adjustedInputs = adjustInputNames(originInputs)
+    val transformer = TensorflowLoader.buildBigDLModel(
+      tfGraph,
+      inputs,
+      endPoints,
+      ByteOrder.LITTLE_ENDIAN,
+      "",
+      Some(context)
+    ).asInstanceOf[Graph[T]]
+
+
+    if (adjustedInputs.nonEmpty) {
+      val inputNodes = originInputs.map(name2Node)
+      val inputDataSeq = inputNodes.map { node => // this is the input op
+        node.element.getOp match {
+          // only support Dequeue before reader
+          case "QueueDequeueV2" => handleLocalDequeue(node, cache)
+        }
+      }
+
+      val reducedInputSeq = inputDataSeq.reduce { (outerSeq1, outerSeq2) =>
+        outerSeq1.zip(outerSeq2).map { case (seq1, seq2) =>
+          seq1.add(seq2)
+        }
+      }
+
+      reducedInputSeq.map { tensors =>
+        val output = transformer.forward(tensors.flatten())
+        toTable(output)
+      }
+    } else {
+      Seq(toTable(transformer.forward(T())))
+    }
+  }
+
+  private def toTable(activity: Activity): Table = {
```

Comment:
why not use the toTable in Activity?

Reply:
if this activity is actually a tensor, toTable will throw an exception